lib/extractor/layer.ts

```diff
@@ -53,6 +55,70 @@ export async function extractDockerArchive(
   });
 }
 
+export async function extractOciArchive(
+  ociArchiveFilesystemPath: string,
+  extractActions: ExtractAction[],
+): Promise<{
+  layers: ExtractedLayers[];
+  manifest: OciArchiveManifest;
+}> {
+  return new Promise((resolve, reject) => {
+    const tarExtractor: Extract = extract();
+
+    const layers: Record<string, ExtractedLayers> = {};
+    const manifests: OciArchiveManifest[] = [];
+    let imageIndex;
+
+    tarExtractor.on("entry", async (header, stream, next) => {
+      if (header.type === "file") {
+        if (isImageIndexFile(header.name)) {
+          imageIndex = await streamToJson<OciImageIndex>(stream);
+        } else {
+          const streamCopy1 = new PassThrough();
+          const streamCopy2 = new PassThrough();
+          stream.pipe(streamCopy1);
+          stream.pipe(streamCopy2);
+
+          const promises = [
+            streamToJson(streamCopy1).catch(() => undefined),
+            extractImageLayer(streamCopy2, extractActions).catch(
+              () => undefined,
+            ),
+          ];
+          const [manifest, layer] = await Promise.all(promises);
+
+          const headerParts = header.name.split("/");
+          const headerName = `sha256:${headerParts[headerParts.length - 1]}`;
+          if (manifest !== undefined) {
```

Comment:
not sure what's my opinion about that you try to parse the file as two options, one throws and one not. why not going with the other option of going by the actual format and understanding which files are layers and which not?

Reply:
>going by the actual format and understanding which files are layers and which not?

Do you mean looking at the manifest to determine this? I think because we have to process the archive 2 times - one to extract the index.json, and another to extract the files we want!

Reply:
I mean, how we've tried it Shai, I couldn't get it working, I don't think you can in any way store streams, you have to process them immediately when you get the event. Maybe there is a way to do it in a similar way like I'm doing it now, just pass callbacks somehow. 
@ivanstanev how we were trying to approach it didn't mean scanning it twice, it was something like that: 
- create an object for streams
- on every tar layer event add stream chunk to the object, with the `header.name` identifier, also we get index.json
- now (further down in the code) we've got `index.json` so we know the manifests and I've tried to loop through the streams and call either manifest extract if it's manifest or layer extract if not. 
Of course that did not work cause streams were finished by that time. But maybe could do something like that with promises 🤔

Reply:
hmmm the downside here is that we parse layers of archs we don't really need. so we parse all the layers, even the ones that are not linux and amd64, and only filter them out in the end

Reply:
But in the other case, we store those streams in the memory, so we could have 2GB worth of streams in the memory, is that not worse?

Reply:
>and only filter them out in the end

You're right but I think it is still better performance-wise than the worst case (when the index.json is located right at the end of the tar-stream) of passing through the stream twice 😕 

Let's look at an example:
![Untitled drawing (5)](https://user-images.githubusercontent.com/16717473/83730011-6290dc80-a640-11ea-9174-4b4d268ec16f.png)

In the current case we process:
- 3GB of data, reading every relevant file of every layer, regardless if it's the wrong OS. The majority of files are _not_ relevant (we have around 8 types that we care about)

In the "use index.json to understand the relevant layers" case we process:
- Almost 6GB of data: almost 3 GB to reach the index.json in order to parse it, and almost 3GB where we skip the first 2GB and process only 1GB for relevant files/layers

Reply:
OK we discussed an alternative that makes more sense, and we can definitely look to improve this in the future so we don't process irrelevant layers